### PR TITLE
allow SffIO and XdnaIO to accept a file path

### PR DESCRIPTION
**Pull request by mdehoon**
This pull request allows SffIO and XdnaIO to write to a file path in addition to a file handle.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqIO/Interfaces.py [hunk 1]
```diff
@@ -94,11 +94,37 @@ class SequenceWriter:
...
+            except TypeError:
+                # target was opened in text mode
+                raise ValueError("File must be opened in binary mode.") from None
+            except AttributeError:
+                # target is a path
+                handle = open(target, mode)
+            else:
+                handle = target
+        else:
+            raise RuntimeError("Unknown mode '%s'" % mode)
```
Comment (peterjc) [inline Bio/SeqIO/Interfaces.py]: ValueError?

File: Bio/SeqIO/SffIO.py [hunk 2]
```diff
@@ -230,10 +230,10 @@
 """
 
 
-from Bio.SeqIO.Interfaces import SequenceWriter
 from Bio import Alphabet
 from Bio.Seq import Seq
 from Bio.SeqRecord import SeqRecord
+from Bio.SeqIO.Interfaces import SequenceWriter
```
Comment (peterjc) [inline Bio/SeqIO/SffIO.py]: If you are sorting alphabetically, it should be one line higher.

See also #2715 when I removed the ``import sys`` and put the standard library imports above the Biopython ones.

File: Bio/SeqIO/SffIO.py [hunk 3]
```diff
@@ -1135,29 +1135,21 @@ def _SffTrimIterator(handle, alphabet=Alphabet.generic_dna):
     """Iterate over SFF reads (as SeqRecord objects) with trimming (PRIVATE)."""
     return SffIterator(handle, alphabet, trim=True)
 
-
```
Comment (peterjc) [inline Bio/SeqIO/SffIO.py]: PEP8 says two lines between classes etc. TravisCI will complain via flake8 or black.

File: Bio/SeqIO/Interfaces.py (hunk 1 above)
Comment (mdehoon) [reply Bio/SeqIO/Interfaces.py]: To me this is a `RuntimeError`. The main (and probably only) purpose of this class is to be used by the other submodules in `Bio.SeqIO`. Then, if the mode is not correct, it means something is wrong in Biopython; it's not a mistake by the user.

Comment (peterjc) [reply Bio/SeqIO/Interfaces.py]: Well OK. I can see a case for either exception.

**Pull request merged**